qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,273 @@
+        self._categories.append(cat)
+
+    def data(self, index, role=Qt.DisplayRole):
+        """Return the item data for index.
+
+        Override QAbstractItemModel::data.
+
+        Args:
+            index: The QModelIndex to get item flags for.
+
+        Return: The item data, or None on an invalid index.
+        """
+        if not index.isValid() or role != Qt.DisplayRole:
+            return
+        if not index.parent().isValid():
+            if index.column() == 0:
+                return self._categories[index.row()].tablename
+        else:
+            table = self._categories[index.parent().row()]
+            idx = table.index(index.row(), index.column())
+            return table.data(idx)
+
+    def flags(self, index):
+        """Return the item flags for index.
+
+        Override QAbstractItemModel::flags.
+
+        Return: The item flags, or Qt.NoItemFlags on error.
+        """
+        if not index.isValid():
+            return
+        if index.parent().isValid():
+            # item
+            return (Qt.ItemIsEnabled | Qt.ItemIsSelectable |
+                    Qt.ItemNeverHasChildren)
+        else:
+            # category
+            return Qt.NoItemFlags
+
+    def index(self, row, col, parent=QModelIndex()):
+        """Get an index into the model.
+
+        Override QAbstractItemModel::index.
+
+        Return: A QModelIndex.
+        """
+        if (row < 0 or row >= self.rowCount(parent) or
+                col < 0 or col >= self.columnCount(parent)):
+            return QModelIndex()
+        if parent.isValid():
+            if parent.column() != 0:
+                return QModelIndex()
+            # store a pointer to the parent table in internalPointer
+            return self.createIndex(row, col, self._categories[parent.row()])
+        return self.createIndex(row, col, None)
+
+    def parent(self, index):
+        """Get an index to the parent of the given index.
+
+        Override QAbstractItemModel::parent.
+
+        Args:
+            index: The QModelIndex to get the parent index for.
+        """
+        parent_table = index.internalPointer()
+        if not parent_table:
+            # categories have no parent
+            return QModelIndex()
+        row = self._categories.index(parent_table)
+        return self.createIndex(row, 0, None)
+
+    def rowCount(self, parent=QModelIndex()):
+        """Override QAbstractItemModel::rowCount."""
+        if not parent.isValid():
+            # top-level
+            return len(self._categories)
+        cat = self._cat_from_idx(parent)
+        if not cat or parent.column() != 0:
+            # item or nonzero category column (only first col has children)
+            return 0
+        else:
+            # category
+            return cat.rowCount()
+
+    def columnCount(self, parent=QModelIndex()):
+        """Override QAbstractItemModel::columnCount."""
+        # pylint: disable=unused-argument
+        return 3
```

Comment:
It doesn't actually seem to care (it actually isn't 3 columns for bookmarks/quickmarks and this still works fine). I remember I had trouble doing it dynamically, but I can try again.